### PR TITLE
fix(GraphQL): Fix mutation with Int Xid variables.

### PR DESCRIPTION
**Pull request by minhaj-shakeel**
Fixes GRAPHQL-1094.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7565)

### CODE REVIEW

**Review by abhimanyusinghgaur — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7565)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @manishrjain, @martinmr, @minhaj-shakeel, @pawanrawal, and @vvbalaji-dgraph)

___
*[graphql/e2e/common/mutation.go, line 4892 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVph5v733KOAsFs632Y:-MVph5v733KOAsFs632Z:b-qs1xre) ([raw file](https://github.com/dgraph-io/dgraph/blob/4607fad6706cf395c4450438df9c060fadbf1821/graphql/e2e/common/mutation.go#L4892)):*
> ```
> $bookId
> ```
so this adds the test for variable but removes the case for hardcoded value.
Let's not do this.

Let's keep the existing one, but just add one more book in the same mutation which uses a variable.
___
*[graphql/e2e/common/mutation.go, line 4925 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVphPv97eS5YFHKNemb:-MVphPvA9uYrmi6XnsYt:b-s93xvo) ([raw file](https://github.com/dgraph-io/dgraph/blob/4607fad6706cf395c4450438df9c060fadbf1821/graphql/e2e/common/mutation.go#L4925)):*
> ```
> $chId
> ```
same comment as the above test
___
*[graphql/resolve/mutation_rewriter.go, line 1758 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVpheVPDMRvOiUzzR4X:-MVpheVPDMRvOiUzzR4Y:bn4nmkx) ([raw file](https://github.com/dgraph-io/dgraph/blob/4607fad6706cf395c4450438df9c060fadbf1821/graphql/resolve/mutation_rewriter.go#L1758)):*
<details><summary><i alt="summary">Quoted 60 lines of code…</i></summary>

> ```
> 					switch xidVal.(type) {
> 					case json.Number:
> 						val, err := xidVal.(json.Number).Int64()
> 						if err != nil {
> 							retErrors = append(retErrors, err)
> 							return nil, retErrors
> 						}
> 						xidString = strconv.FormatInt(val, 10)
> 					default:
> 						val, ok := xidVal.(int64)
> 						if !ok {
> 							retErrors = append(retErrors, errors.New(fmt.Sprintf("encountered an XID %s with %s that isn't "+
> 								"a Int but data type in schema is Int", xid.Name(), xid.Type().Name())))
> 							return nil, retErrors
> 						}
> 						xidString = strconv.FormatInt(val, 10)
> 					}
> 				case "Int64":
> 					switch xidVal.(type) {
> 					case json.Number:
> 						val, err := xidVal.(json.Number).Int64()
> 						if err != nil {
> 							retErrors = append(retErrors, err)
> 							return nil, retErrors
> 						}
> 						xidString = strconv.FormatInt(val, 10)
> 					case int64:
> 						val, ok := xidVal.(int64)
> 						if !ok {
> 							retErrors = append(retErrors, errors.New(fmt.Sprintf("encountered an XID %s with %s that isn't "+
> 								"a Int64 but data type in schema is Int64", xid.Name(), xid.Type().Name())))
> 							return nil, retErrors
> 						}
> 						xidString = strconv.FormatInt(val, 10)
> 					default:
> 						xidString, ok = xidVal.(string)
> 						if !ok {
> 							retErrors = append(retErrors, errors.New(fmt.Sprintf("encountered an XID %s with %s that isn't "+
> 								"a Int64", xid.Name(), xid.Type().Name())))
> 							return nil, retErrors
> 						}
> 					}
> 				case "Float":
> 					switch xidVal.(type) {
> 					case json.Number:
> 						val, err := xidVal.(json.Number).Float64()
> 						if err != nil {
> 							retErrors = append(retErrors, err)
> 							return nil, retErrors
> 						}
> 						xidString = strconv.FormatFloat(val, 'f', -1, 64)
> 					default:
> 						val, ok := xidVal.(float64)
> 						if !ok {
> 							retErrors = append(retErrors, errors.New(fmt.Sprintf("encountered an XID %s with %s that isn't "+
> 								"a Float but data type in schema is Float", xid.Name(), xid.Type().Name())))
> 							return nil, retErrors
> 						}
> 						xidString = strconv.FormatFloat(val, 'f', -1, 64)
> 					}
> ```
</details>

so this block of code seems repetitive.
So, put this in a function
```
func extractVal(xidVal interface{}, typename string) (string, error)
```
and use at both the places.

**Review by abhimanyusinghgaur — COMMENTED** [review]
This should go in `release/v21.03` not `master`.

File: graphql/resolve/mutation_rewriter.go [hunk 1]
```diff
@@ -1383,11 +1383,34 @@ func rewriteObject(
 				// TODO: Add a function for parsing idVal. This is repeatitive
```
Comment (vmrajas) [inline graphql/resolve/mutation_rewriter.go:1383]: You may remove this TODO after addressing Abhimanyu's comment on writing extractVal function.

File: graphql/resolve/mutation_rewriter.go [hunk 2]
```diff
@@ -1383,11 +1383,34 @@ func rewriteObject(
...
+					default:
+						val, _ := xidVal.(int64)
+						xidString = strconv.FormatInt(val, 10)
+					}
+				case "Int64":
+					switch xidVal.(type) {
+					case json.Number:
+						val, _ := xidVal.(json.Number).Int64()
+						xidString = strconv.FormatInt(val, 10)
+					case int64:
```
Comment (vmrajas) [inline graphql/resolve/mutation_rewriter.go:1399]: This part is different for int64 than for int or float. Add a comment here on why that is so and the fact that default handles the case for string.

**Review by abhimanyusinghgaur — CHANGES_REQUESTED** [review]
We should close this PR, and raise a new one against the `release/v21.03` branch with the changes in this.

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-:-MVtzdIi3CMCjGilMrrK:b-vcm3ub)* status: 0 of 2 files reviewed, 10 unresolved discussions (waiting on @manishrjain, @martinmr, @minhaj-shakeel, @pawanrawal, and @vvbalaji-dgraph)

___
*[graphql/resolve/mutation_rewriter.go, line 2264 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVtwmGnAxkNpjIvX3WL:-MVtwmGnAxkNpjIvX3WM:bdx49z6) ([raw file](https://github.com/dgraph-io/dgraph/blob/f287df5ce070855a62211c9366555ffb99171b2b/graphql/resolve/mutation_rewriter.go#L2264)):*
> ```
> switch xidVal.(type)
> ```
```
switch xVal := xidVal.(type)
```
And then use `xVal`, no need to again cast it, if we are already using a type switch.
Also, as we are using a type switch, we can just have explicit cases for `json.Number` and `int64`, then the default case should return errors only. 
At present, the default case handles the `int64` case too, which is a bit ugly.

Same comments for other `Int64` and `Float` cases.
___
*[graphql/resolve/mutation_rewriter.go, line 2288 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVtxu8A0GcqhJopsEOh:-MVtxu8A0GcqhJopsEOi:b-xo5861) ([raw file](https://github.com/dgraph-io/dgraph/blob/f287df5ce070855a62211c9366555ffb99171b2b/graphql/resolve/mutation_rewriter.go#L2288)):*
> ```
> .(int64)
> ```
this shouldn't be required after addressing the above comments, as we are already in the case for `int64`.
___
*[graphql/resolve/mutation_rewriter.go, line 2294 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVtykADDMcLkjsWsn6H:-MVtykADDMcLkjsWsn6I:b-jukjfz) ([raw file](https://github.com/dgraph-io/dgraph/blob/f287df5ce070855a62211c9366555ffb99171b2b/graphql/resolve/mutation_rewriter.go#L2294)):*
> ```
>  as per spec
> ```
remove this bit.
It is not as per spec, it is just what we allow for convenience to clients.
___
*[graphql/resolve/mutation_rewriter.go, line 2295 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVtyw1W6YyO6l3ZNSiA:-MVtyw1W6YyO6l3ZNSiB:bp8d41) ([raw file](https://github.com/dgraph-io/dgraph/blob/f287df5ce070855a62211c9366555ffb99171b2b/graphql/resolve/mutation_rewriter.go#L2295)):*
> ```
>  id
> ```
typo here?
___
*[graphql/resolve/mutation_rewriter.go, line 2321 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVtzH5r7gzM2tiV-Nkg:-MVtzH5r7gzM2tiV-Nkh:b-tvg4k7) ([raw file](https://github.com/dgraph-io/dgraph/blob/f287df5ce070855a62211c9366555ffb99171b2b/graphql/resolve/mutation_rewriter.go#L2321)):*
<details><summary><i alt="summary">Quoted 7 lines of code…</i></summary>

> ```
> 	default:
> 		xidString, ok := xidVal.(string)
> 		if !ok {
> 			return "", errors.New(fmt.Sprintf("encountered an XID %s with %s that isn't "+
> 				"a String", xidName, typeName))
> 		}
> 		return xidString, nil
> ```
</details>

we should explicitly have a `case "String"` that handles this.
The `default` case then can just always return an error that the type is not allowed as xid.

**Review by minhaj-shakeel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7565)* status: 0 of 2 files reviewed, 10 unresolved discussions (waiting on @abhimanyusinghgaur, @manishrjain, @martinmr, @pawanrawal, @vmrajas, and @vvbalaji-dgraph)

___
*[graphql/e2e/common/mutation.go, line 4892 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVph5v733KOAsFs632Y:-MVun-HL6lDNJ3ZSRAOY:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4607fad6706cf395c4450438df9c060fadbf1821/graphql/e2e/common/mutation.go#L4892)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
> $bookId
> ```
so this adds the test for variable but removes the case for hardcoded value.
Let's not do this.

Let's keep the existing one, but just add one more book in the same mutation which uses a variable.
</blockquote></details>

Done.
___
*[graphql/e2e/common/mutation.go, line 4925 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVphPv97eS5YFHKNemb:-MVumzsQ4SRM4XHR3rS-:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4607fad6706cf395c4450438df9c060fadbf1821/graphql/e2e/common/mutation.go#L4925)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
> $chId
> ```
same comment as the above test
</blockquote></details>

Done.
___
*[graphql/resolve/mutation_rewriter.go, line 1383 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVoal__Sz0Hx3PJyD35-r1-1383:-MVun06n3u-qaq6hAg0Y:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4607fad6706cf395c4450438df9c060fadbf1821/graphql/resolve/mutation_rewriter.go#L1383)):*
<details><summary><i>Previously, vmrajas wrote…</i></summary><blockquote>

You may remove this TODO after addressing Abhimanyu's comment on writing extractVal function.
</blockquote></details>

Done.
___
*[graphql/resolve/mutation_rewriter.go, line 1399 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVoal__Sz0Hx3PJyD35-r1-1399:-MVun0Qg4rLYj8sPSsKv:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4607fad6706cf395c4450438df9c060fadbf1821/graphql/resolve/mutation_rewriter.go#L1399)):*
<details><summary><i>Previously, vmrajas wrote…</i></summary><blockquote>

This part is different for int64 than for int or float. Add a comment here on why that is so and the fact that default handles the case for string.
</blockquote></details>

Done.
___
*[graphql/resolve/mutation_rewriter.go, line 1758 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVpheVPDMRvOiUzzR4X:-MVun2Ke1QX06-Ygeb57:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/4607fad6706cf395c4450438df9c060fadbf1821/graphql/resolve/mutation_rewriter.go#L1758)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
> 					switch xidVal.(type) {
> 					case json.Number:
> 						val, err := xidVal.(json.Number).Int64()
> 						if err != nil {
> 							retErrors = append(retErrors, err)
> 							return nil, retErrors
> 						}
> 						xidString = strconv.FormatInt(val, 10)
> 					default:
> 						val, ok := xidVal.(int64)
> 						if !ok {
> 							retErrors = append(retErrors, errors.New(fmt.Sprintf("encountered an XID %s with %s that isn't "+
> 								"a Int but data type in schema is Int", xid.Name(), xid.Type().Name())))
> 							return nil, retErrors
> 						}
> 						xidString = strconv.FormatInt(val, 10)
> 					}
> 				case "Int64":
> 					switch xidVal.(type) {
> 					case json.Number:
> 						val, err := xidVal.(json.Number).Int64()
> 						if err != nil {
> 							retErrors = append(retErrors, err)
> 							return nil, retErrors
> 						}
> 						xidString = strconv.FormatInt(val, 10)
> 					case int64:
> 						val, ok := xidVal.(int64)
> 						if !ok {
> 							retErrors = append(retErrors, errors.New(fmt.Sprintf("encountered an XID %s with %s that isn't "+
> 								"a Int64 but data type in schema is Int64", xid.Name(), xid.Type().Name())))
> 							return nil, retErrors
> 						}
> 						xidString = strconv.FormatInt(val, 10)
> 					default:
> 						xidString, ok = xidVal.(string)
> 						if !ok {
> 							retErrors = append(retErrors, errors.New(fmt.Sprintf("encountered an XID %s with %s that isn't "+
> 								"a Int64", xid.Name(), xid.Type().Name())))
> 							return nil, retErrors
> 						}
> 					}
> 				case "Float":
> 					switch xidVal.(type) {
> 					case json.Number:
> 						val, err := xidVal.(json.Number).Float64()
> 						if err != nil {
> 							retErrors = append(retErrors, err)
> 							return nil, retErrors
> 						}
> 						xidString = strconv.FormatFloat(val, 'f', -1, 64)
> 					default:
> 						val, ok := xidVal.(float64)
> 						if !ok {
> 							retErrors = append(retErrors, errors.New(fmt.Sprintf("encountered an XID %s with %s that isn't "+
> 								"a Float but data type in schema is Float", xid.Name(), xid.Type().Name())))
> 							return nil, retErrors
> 						}
> 						xidString = strconv.FormatFloat(val, 'f', -1, 64)
> 					}
> ```
so this block of code seems repetitive.
So, put this in a function
```
func extractVal(xidVal interface{}, typename string) (string, error)
```
and use at both the places.
</blockquote></details>

Done.
___
*[graphql/resolve/mutation_rewriter.go, line 2264 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVtwmGnAxkNpjIvX3WL:-MVun2xS44MXExUXKQ6Q:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f287df5ce070855a62211c9366555ffb99171b2b/graphql/resolve/mutation_rewriter.go#L2264)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
> switch xidVal.(type)
> ```
```
switch xVal := xidVal.(type)
```
And then use `xVal`, no need to again cast it, if we are already using a type switch.
Also, as we are using a type switch, we can just have explicit cases for `json.Number` and `int64`, then the default case should return errors only. 
At present, the default case handles the `int64` case too, which is a bit ugly.

Same comments for other `Int64` and `Float` cases.
</blockquote></details>

Done.
___
*[graphql/resolve/mutation_rewriter.go, line 2288 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVtxu8A0GcqhJopsEOh:-MVun3WeA5R2SSLFZGex:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f287df5ce070855a62211c9366555ffb99171b2b/graphql/resolve/mutation_rewriter.go#L2288)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
> .(int64)
> ```
this shouldn't be required after addressing the above comments, as we are already in the case for `int64`.
</blockquote></details>

Done.
___
*[graphql/resolve/mutation_rewriter.go, line 2294 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVtykADDMcLkjsWsn6H:-MVun4T-Ad6gBOlM4RZ2:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f287df5ce070855a62211c9366555ffb99171b2b/graphql/resolve/mutation_rewriter.go#L2294)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
>  as per spec
> ```
remove this bit.
It is not as per spec, it is just what we allow for convenience to clients.
</blockquote></details>

Done.
___
*[graphql/resolve/mutation_rewriter.go, line 2295 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVtyw1W6YyO6l3ZNSiA:-MVun4kY7bHHinWDaVQs:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f287df5ce070855a62211c9366555ffb99171b2b/graphql/resolve/mutation_rewriter.go#L2295)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
>  id
> ```
typo here?
</blockquote></details>

Done.
___
*[graphql/resolve/mutation_rewriter.go, line 2321 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-MVtzH5r7gzM2tiV-Nkg:-MVun5Z32UYb7qZrS0DI:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f287df5ce070855a62211c9366555ffb99171b2b/graphql/resolve/mutation_rewriter.go#L2321)):*
<details><summary><i>Previously, abhimanyusinghgaur (Abhimanyu Singh Gaur) wrote…</i></summary><blockquote>

> ```
> 	default:
> 		xidString, ok := xidVal.(string)
> 		if !ok {
> 			return "", errors.New(fmt.Sprintf("encountered an XID %s with %s that isn't "+
> 				"a String", xidName, typeName))
> 		}
> 		return xidString, nil
> ```
we should explicitly have a `case "String"` that handles this.
The `default` case then can just always return an error that the type is not allowed as xid.
</blockquote></details>

Done.

**Review by abhimanyusinghgaur — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7565#-:-MVuphR_AKA3tuikwUUE:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @abhimanyusinghgaur, @manishrjain, @martinmr, @pawanrawal, @vmrajas, and @vvbalaji-dgraph)

**Pull request merged**